### PR TITLE
Position Control Refactoring Part 4

### PR DESCRIPTION
**Pull request by MaEtUgR**
**Describe problem solved by this pull request**
This is part 4 of the refactoring from the monster pr #12072 and should be easy to understand and review since the diff is manageable.

**Describe your solution**
- Got rid of the ModuelParam dependency of PositionControl and instead set gains and limits via setters that have a clear interface description
- Added unit tests that mostly test sanity of the controller e.g. it can take all sorts of setpoint combinations, goes to the right direction, respects the limits

**Test data / coverage**
SITL hover goto tested and unit tests pass.

**Additional context**
Part 3 that goes before this: #13262

### CODE REVIEW

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 1]
```diff
@@ -341,6 +352,15 @@ MulticopterPositionControl::parameters_update(bool force)
 		ModuleParams::updateParams();
 		SuperBlock::updateParams();
 
+		_control.setPositionGains(Vector3f(_param_mpc_xy_p.get(), _param_mpc_xy_p.get(), _param_mpc_z_p.get()));
+		_control.setVelocityGains(Vector3f(_param_mpc_xy_vel_p.get(), _param_mpc_xy_vel_p.get(), _param_mpc_z_vel_p.get()),
+					  Vector3f(_param_mpc_xy_vel_i.get(), _param_mpc_xy_vel_i.get(), _param_mpc_z_vel_i.get()),
+					  Vector3f(_param_mpc_xy_vel_d.get(), _param_mpc_xy_vel_d.get(), _param_mpc_z_vel_d.get()));
+		_control.setVelocityLimits(_param_mpc_xy_vel_max.get(), _param_mpc_z_vel_max_up.get(), _param_mpc_z_vel_max_dn.get());
+		_control.setThrustLimits(_param_mpc_thr_min.get(), _param_mpc_thr_max.get());
+		_control.setTiltLimit(M_DEG_TO_RAD_F * _param_mpc_tiltmax_air.get()); // convert to radians!
```
Comment (dagar) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]:
```suggestion
		_control.setTiltLimit(math::radians(_param_mpc_tiltmax_air.get())); // convert to radians!
```

Pretty minor, but I think we're pretty close to not needing these defines at all.

Comment (MaEtUgR) [reply src/modules/mc_pos_control/mc_pos_control_main.cpp]: Ouch, I missed you trying to replace them. Good hint, I'll follow your steps 👍

**Comment by MaEtUgR** [comment]
I rebased and corrected according to @dagar's comment.

**Comment by LorenzMeier** [comment]
@MaEtUgR What is missing to get this in?

**Comment by bresch** [comment]
@LorenzMeier My review. I'll allocate some time tomorrow to do that.

File: src/modules/mc_pos_control/PositionControl/PositionControl.hpp [hunk 2]
```diff
@@ -181,42 +209,40 @@ class PositionControl : public ModuleParams
...
-	matrix::Vector3f _acc_sp{}; /**< desired acceleration: not supported yet */
-	matrix::Vector3f _thr_sp{}; /**< desired thrust */
+	// Gains
+	matrix::Vector3f _gain_pos_p; ///< Position control proportional gain
+	matrix::Vector3f _gain_vel_p; ///< Velocity control proportional gain
+	matrix::Vector3f _gain_vel_i; ///< Velocity control integral gain
+	matrix::Vector3f _gain_vel_d; ///< Velocity control derivative gain
+
+	// Limits
+	float _lim_vel_horizontal{0}; ///< Horizontal velocity limit with feed forward and position control
```
Comment (bresch) [inline src/modules/mc_pos_control/PositionControl/PositionControl.hpp]: For a float, I would rather put `{0.f}` or `{}`

File: src/modules/mc_pos_control/PositionControl/PositionControl.hpp [hunk 3]
```diff
@@ -181,42 +209,40 @@ class PositionControl : public ModuleParams
...
+	float _lim_tilt{0}; ///< Maximum tilt from level the output attitude is allowed to have
+
+	float _hover_thrust{0}; ///< Thrust [0,1] with which the vehicle hovers not aacelerating down or up with level orientation
+
+	// States
+	matrix::Vector3f _pos; /**< position */
+	matrix::Vector3f _vel; /**< velocity */
+	matrix::Vector3f _vel_dot; /**< velocity derivative (replacement for acceleration estimate) */
+	matrix::Vector3f _thr_int; /**< integral term of the velocity controller */
+	float _yaw = 0.0f; /**< yaw */
```
Comment (bresch) [inline src/modules/mc_pos_control/PositionControl/PositionControl.hpp]:
```suggestion
	float _yaw{}; /**< yaw */
```

Comment (bresch) [reply src/modules/mc_pos_control/PositionControl/PositionControl.hpp]: Also, is the comment really useful?

File: src/modules/mc_pos_control/PositionControl/PositionControlTest.cpp [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+	EXPECT_EQ(attitude.fw_control_yaw, false);
+	EXPECT_EQ(attitude.apply_flaps, 0);//vehicle_attitude_setpoint_s::FLAPS_OFF); // TODO why no reference?
+}
+
+class PositionControlBasicTest : public ::testing::Test
+{
+public:
+	PositionControlBasicTest()
+	{
+		_position_control.setPositionGains(Vector3f(1, 1, 1));
```
Comment (bresch) [inline src/modules/mc_pos_control/PositionControl/PositionControlTest.cpp]:
```suggestion
		_position_control.setPositionGains(Vector3f(1.f, 1.f, 1.f));
```

File: src/modules/mc_pos_control/PositionControl/PositionControl.hpp (hunk 3 above)
Comment (MaEtUgR) [reply src/modules/mc_pos_control/PositionControl/PositionControl.hpp]: I also changed the comment to be useful.

**Comment by MaEtUgR** [comment]
@bresch Thanks for the review! I rebased on master and fixed all the comments. Now it should be consistent with the initializations.

**Pull request merged**